src/commcare_cloud/commands/terraform/aws.py

```diff
@@ -287,13 +289,108 @@ class AwsSignIn(CommandBase):
-        aws_profile = environment.terraform_config.aws_profile
-        aws_sign_in(aws_profile, duration_minutes, force_new=True)
+        aws_sign_in(environment, duration_minutes, force_new=True)
 
 
 @memoized
-def aws_sign_in(aws_profile, duration_minutes=DEFAULT_SIGN_IN_DURATION_MINUTES,
-                force_new=False):
+def aws_sign_in(environment, duration_minutes=DEFAULT_SIGN_IN_DURATION_MINUTES, force_new=False):
+    if environment.aws_config.credential_style == 'sso':
+        return _aws_sign_in_with_sso(environment)
+    else:
+        return _aws_sign_in_with_iam(environment.terraform_config.aws_profile, duration_minutes=duration_minutes,
+                                     force_new=force_new)
+
+
+@memoized
+def _aws_sign_in_with_sso(environment):
+    """
+    Create a temp session through MFA for a given aws profile
+
+    :param aws_profile: The name of an existing aws profile to create a temp session for
+    :param duration_minutes: How long to set the session expiration if a new one is created
+    :param force_new: If set to True, creates new credentials even if valid ones are found
+    :return: The name of temp session profile.
+             (Always the passed in profile followed by ':session')
+    """
+    aws_session_profile = '{}:session'.format(environment.terraform_config.aws_profile)
+    # todo: add `... or if _date_modified(aws_config_path) > _date_modified(aws_credentials_path)`
+    if not _has_profile_for_sso(aws_session_profile):
+        puts(color_notice("Configuring SSO. To further customize, run `aws configure sso --profile {}`".format(aws_session_profile)))
+        _write_profile_for_sso(
+            aws_session_profile,
+            sso_start_url=environment.aws_config.sso_config.sso_start_url,
+            sso_account_id=environment.aws_config.sso_config.sso_account_id,
+            sso_region=environment.aws_config.sso_config.sso_region,
+            region=environment.aws_config.sso_config.region,
+        )
+
+    if not _has_valid_session_credentials_for_sso():
+        check_output(['aws', 'sso', 'login'], env={'AWS_PROFILE': aws_session_profile})
+
+    if not _has_valid_v1_session_credentials(aws_session_profile):
+        caller_identity = _get_caller_identity({'AWS_PROFILE': aws_session_profile})
+        if not caller_identity:
+            raise ValueError((
+                "SSO profile mis-configured and cannot fix automatically. "
+                "Edit [profile {}] in ~/.aws/config manually;"
+                "to start over, remove it from the file manually and try again.").format(aws_session_profile))
+        # Until this is built in to aws, we need this insane workaround to get backwards compatibility
+        # with the credential format that terraform uses
+        cli_cache_dir = os.path.expanduser('~/.aws/cli/cache/')
+
+        for filename in os.listdir(cli_cache_dir):
+            with open(os.path.join(cli_cache_dir, filename)) as f:
```

Comment:
This assumes that every item in `cli_cache_dir` can be opened as a file, but I think this will fail if there are directories in there. What do you think of using `os.path.isfile()` to check before opening?

Reply:
0a0a119